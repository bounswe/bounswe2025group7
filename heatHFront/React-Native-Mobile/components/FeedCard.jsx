import React from 'react';
import { View, Text, Image, StyleSheet } from 'react-native';

const FeedCard = ({ feed }) => {
  // Use recipe photo as fallback if feed.image is null
  const imageUrl = feed.image || feed.recipe?.photo;

  return (
    <View style={styles.card}>
      <View style={styles.header}>
        <Image source={{ uri: feed.profilePhoto }} style={styles.avatar} />
        <View style={styles.nameContainer}>
          <Text style={styles.name}>{feed.name} {feed.surname}</Text>
          {/* Wrapping the timestamp in <Text> */}
          <Text style={styles.timestamp}>{feed.createdAt}</Text>
        </View>
      </View>

      {/* Wrapping the text in <Text> */}
      <Text style={styles.text}>{feed.text}</Text>

      {/* Conditionally rendering the image or recipe photo */}
      {imageUrl && <Image source={{ uri: imageUrl }} style={styles.feedImage} />}
      
      {/* Rendering the recipe only if it exists */}
      {feed.recipe && (
        <View style={styles.recipeContainer}>
          <Text style={styles.recipeTitle}>üçΩ {feed.recipe.title}</Text>
          <Text style={styles.savedText}>{feed.savedByCurrentUser ? 'Saved' : 'Not Saved'}</Text>
        </View>
      )}

      {/* Rendering the footer with likes and comments */}
      <View style={styles.footer}>
        <Text style={styles.footerText}>‚ù§Ô∏è {feed.likeCount} | üí¨ {feed.commentCount}</Text>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  card: {
    padding: 16,
    margin: 12,
    backgroundColor: '#fff',
    borderRadius: 16,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 4 },
    shadowOpacity: 0.1,
    shadowRadius: 8,
    elevation: 5,  // For Android shadow
    overflow: 'hidden', // To ensure rounded corners for images
  },
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 12,
  },
  avatar: {
    width: 45,
    height: 45,
    borderRadius: 22.5,
    marginRight: 12,
    borderWidth: 1,
    borderColor: '#ddd',  // Light border around avatar
  },
  nameContainer: {
    justifyContent: 'center',
  },
  name: {
    fontWeight: 'bold',
    fontSize: 16,
    color: '#333',
  },
  timestamp: {
    fontSize: 12,
    color: '#999',
  },
  text: {
    fontSize: 14,
    color: '#555',
    lineHeight: 20,
    marginBottom: 12,
  },
  feedImage: {
    width: '100%',
    height: 250,
    marginTop: 12,
    borderRadius: 10,
    borderWidth: 1,
    borderColor: '#eee',
  },
  recipeContainer: {
    marginTop: 12,
    padding: 10,
    backgroundColor: '#f7f7f7',
    borderRadius: 8,
    marginBottom: 12,
  },
  recipeTitle: {
    fontWeight: 'bold',
    color: '#333',
  },
  savedText: {
    fontSize: 12,
    color: '#777',
  },
  footer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 12,
  },
  footerText: {
    fontSize: 14,
    color: '#555',
    fontWeight: '600',
  },
});

export default FeedCard;
