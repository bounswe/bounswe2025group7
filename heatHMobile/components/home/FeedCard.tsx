import { useMemo, useState } from 'react';
import { View, Text, Image, Pressable } from 'react-native';
import { ThemedText } from '@/components/themed-text';
import { ThemedView } from '@/components/themed-view';
import Avatar from '@/components/user/Avatar';
import Button from '@/components/ui/Button';
import { FeedResponse, feedService } from '@/services/feedService';

export default function FeedCard({ feed }: { feed: FeedResponse }) {
  const [liked, setLiked] = useState<boolean>(feed.likedByCurrentUser);
  const [likeCount, setLikeCount] = useState<number>(feed.likeCount);

  const handleToggleLike = async () => {
    try {
      const nextLiked = !liked;
      setLiked(nextLiked);
      setLikeCount((c) => (nextLiked ? c + 1 : Math.max(0, c - 1)));
      if (nextLiked) {
        await feedService.likeFeed(feed.id);
      } else {
        await feedService.unlikeFeed(feed.id);
      }
    } catch (e) {
      setLiked(feed.likedByCurrentUser);
      setLikeCount(feed.likeCount);
    }
  };

  const headerName = useMemo(() => {
    const name = [feed.name, feed.surname].filter(Boolean).join(' ').trim();
    return name.length > 0 ? name : 'User';
  }, [feed.name, feed.surname]);

  return (
    <ThemedView style={{ borderWidth: 1, borderColor: '#e5e7eb', borderRadius: 12, padding: 12, marginBottom: 12 }}>
      <View style={{ flexDirection: 'row', alignItems: 'center', marginBottom: 8 }}>
        <Avatar uri={feed.profilePhoto} size={40} />
        <View style={{ marginLeft: 10 }}>
          <ThemedText style={{ fontWeight: '600' }}>{headerName}</ThemedText>
          <ThemedText style={{ fontSize: 12, opacity: 0.7 }}>{new Date(feed.createdAt).toLocaleString()}</ThemedText>
        </View>
      </View>

      {feed.text ? <ThemedText style={{ marginBottom: 8 }}>{feed.text}</ThemedText> : null}

      {feed.image ? (
        <Image source={{ uri: feed.image }} style={{ width: '100%', height: 220, borderRadius: 10, backgroundColor: '#f3f4f6' }} />
      ) : null}

      {feed.type === 'RECIPE' && feed.recipe ? (
        <View style={{ marginTop: 8, padding: 10, borderRadius: 10, backgroundColor: '#f9fafb' }}>
          <ThemedText style={{ fontWeight: '600', marginBottom: 4 }}>{feed.recipe.title || 'Recipe'}</ThemedText>
          {feed.recipe.photo ? (
            <Image source={{ uri: feed.recipe.photo }} style={{ width: '100%', height: 180, borderRadius: 8, backgroundColor: '#f3f4f6' }} />
          ) : null}
        </View>
      ) : null}

      <View style={{ flexDirection: 'row', alignItems: 'center', marginTop: 10, justifyContent: 'space-between' }}>
        <Pressable onPress={handleToggleLike} style={{ paddingVertical: 6, paddingHorizontal: 10, borderRadius: 8, backgroundColor: liked ? '#0a7ea4' : '#e5e7eb' }}>
          <Text style={{ color: liked ? '#ffffff' : '#111827' }}>{liked ? 'Liked' : 'Like'} · {likeCount}</Text>
        </Pressable>
        <ThemedText style={{ opacity: 0.7 }}>Comments · {feed.commentCount}</ThemedText>
      </View>
    </ThemedView>
  );
}


