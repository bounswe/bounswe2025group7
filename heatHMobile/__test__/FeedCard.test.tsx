import React from 'react';
import { render, screen } from '@testing-library/react-native';
import FeedCard from '@/components/feedCard';

const baseFeed = {
  id: 1,
  userId: 1,
  type: 'TEXT' as const,
  text: 'Hello world',
  likeCount: 5,
  commentCount: 2,
  name: 'Jane',
  surname: 'Doe',
  profilePhoto: null,
};

describe('FeedCard', () => {
  it('renders name, time (if provided), text and meta', () => {
    const feed = { ...baseFeed, createdAt: '2024-01-01T10:00:00.000Z' };
    render(<FeedCard feed={feed as any} />);

    expect(screen.getByText('Jane Doe')).toBeTruthy();
    expect(screen.getByText('Hello world')).toBeTruthy();
    expect(screen.getByText(/❤️ 5/)).toBeTruthy();
    expect(screen.getByText(/💬 2/)).toBeTruthy();
    expect(screen.getByText(/#TEXT/)).toBeTruthy();
  });

  it('shows image for IMAGE_AND_TEXT and recipe photo for RECIPE', () => {
    const imageFeed = { ...baseFeed, type: 'IMAGE_AND_TEXT' as const, image: 'http://example.com/img.jpg' };
    const recipeFeed = { ...baseFeed, type: 'RECIPE' as const, recipe: { photo: 'http://example.com/photo.jpg' } };

    const { rerender, toJSON } = render(<FeedCard feed={imageFeed as any} />);
    expect(toJSON()).toMatchSnapshot();

    rerender(<FeedCard feed={recipeFeed as any} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('falls back to "User" when no name/surname', () => {
    const feed = { ...baseFeed, name: undefined, surname: undefined };
    render(<FeedCard feed={feed as any} />);
    expect(screen.getByText('User')).toBeTruthy();
  });
});


