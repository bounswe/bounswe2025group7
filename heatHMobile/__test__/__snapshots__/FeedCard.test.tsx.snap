// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedCard shows image for IMAGE_AND_TEXT and recipe photo for RECIPE 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "borderColor": "#e5e7eb",
      "borderRadius": 12,
      "borderWidth": 1,
      "marginBottom": 12,
      "padding": 12,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 8,
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#e5e7eb",
            "borderRadius": 20,
            "height": 40,
            "width": 40,
          },
          {
            "borderColor": "#d1d5db",
            "borderWidth": 1,
          },
        ]
      }
    />
    <View
      style={
        {
          "flex": 1,
          "marginLeft": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "#111827",
            "fontWeight": "600",
          }
        }
      >
        Jane Doe
      </Text>
    </View>
  </View>
  <Text
    style={
      {
        "color": "#111827",
        "marginBottom": 8,
      }
    }
  >
    Hello world
  </Text>
  <Image
    resizeMode="cover"
    source={
      {
        "uri": "http://example.com/img.jpg",
      }
    }
    style={
      {
        "backgroundColor": "#ffffff",
        "borderRadius": 8,
        "height": 200,
        "marginBottom": 8,
        "width": "100%",
      }
    }
  />
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        {
          "color": "#4b5563",
          "fontSize": 12,
        }
      }
    >
      ‚ù§Ô∏è 
      5
    </Text>
    <Text
      style={
        {
          "color": "#4b5563",
          "fontSize": 12,
        }
      }
    >
      üí¨ 
      2
    </Text>
    <Text
      style={
        {
          "color": "#4b5563",
          "fontSize": 12,
        }
      }
    >
      #
      IMAGE_AND_TEXT
    </Text>
  </View>
</View>
`;

exports[`FeedCard shows image for IMAGE_AND_TEXT and recipe photo for RECIPE 2`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "borderColor": "#e5e7eb",
      "borderRadius": 12,
      "borderWidth": 1,
      "marginBottom": 12,
      "padding": 12,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 8,
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#e5e7eb",
            "borderRadius": 20,
            "height": 40,
            "width": 40,
          },
          {
            "borderColor": "#d1d5db",
            "borderWidth": 1,
          },
        ]
      }
    />
    <View
      style={
        {
          "flex": 1,
          "marginLeft": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "#111827",
            "fontWeight": "600",
          }
        }
      >
        Jane Doe
      </Text>
    </View>
  </View>
  <Text
    style={
      {
        "color": "#111827",
        "marginBottom": 8,
      }
    }
  >
    Hello world
  </Text>
  <Image
    resizeMode="cover"
    source={
      {
        "uri": "http://example.com/photo.jpg",
      }
    }
    style={
      {
        "backgroundColor": "#ffffff",
        "borderRadius": 8,
        "height": 200,
        "marginBottom": 8,
        "width": "100%",
      }
    }
  />
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        {
          "color": "#4b5563",
          "fontSize": 12,
        }
      }
    >
      ‚ù§Ô∏è 
      5
    </Text>
    <Text
      style={
        {
          "color": "#4b5563",
          "fontSize": 12,
        }
      }
    >
      üí¨ 
      2
    </Text>
    <Text
      style={
        {
          "color": "#4b5563",
          "fontSize": 12,
        }
      }
    >
      #
      RECIPE
    </Text>
  </View>
</View>
`;


